cabal-install/Distribution/Client/Setup.hs

```diff
@@ -387,18 +399,96 @@ instance Monoid GlobalFlags where
     where combine field = field a `mappend` field b
 
 withGlobalRepos :: Verbosity -> GlobalFlags -> ([Repo] -> IO a) -> IO a
-withGlobalRepos _verbosity globalFlags callback =
-    callback $ remoteRepos ++ localRepos
+withGlobalRepos verbosity globalFlags callback =
+    withSecureRepos verbosity secureRepos $ \secureRepos' ->
+      callback $ secureRepos' ++ remoteRepos ++ localRepos
   where
+    secureRepos =
+      [ (remote, cacheDir)
+      | remote <- fromNubList $ globalRemoteRepos globalFlags
+      , remoteRepoSecure remote
+      , let cacheDir = fromFlag (globalCacheDir globalFlags)
+                   </> remoteRepoName remote ]
     remoteRepos =
       [ RepoRemote remote cacheDir
       | remote <- fromNubList $ globalRemoteRepos globalFlags
+      , not (remoteRepoSecure remote)
       , let cacheDir = fromFlag (globalCacheDir globalFlags)
                    </> remoteRepoName remote ]
     localRepos =
       [ RepoLocal local
       | local <- fromNubList $ globalLocalRepos globalFlags ]
 
+-- | Initialize the provided secure repositories
+--
+-- Assumed invariant: `remoteRepoSecure` should be set for all these repos.
+withSecureRepos :: forall a.
+                   Verbosity
+                -> [(RemoteRepo, FilePath)]  -- ^ Repos and cacheDirs
+                -> ([Repo] -> IO a)          -- ^ Callback
+                -> IO a
+withSecureRepos verbosity repos callback = go [] repos
+  where
+    go :: [Repo] -> [(RemoteRepo, FilePath)] -> IO a
+    go acc []                = callback (reverse acc)
+    go acc ((r,cacheDir):rs) = do
+      cachePath <- Sec.makeAbsolute $ Sec.fromFilePath cacheDir
+      withSecureRepo verbosity r cachePath $ \r' ->
+        go (RepoSecure r cacheDir r':acc) rs
+
+-- | Initialize the given secure repo
+--
+-- The security library has its own concept of a "local" repository, distinct
```

Comment:
> Cabal's concept of a local repository is quite different from any other kind of repository

Do these differences really affect users that much? I think our users can live with the fact that secure and non-secure local repos have different on-disk format (provided that the error message printed when the `local-repo` is in the wrong format is decent). IMO, `local-repo` is a more logical/discoverable place to specify secure local repos in the config file than `remote-repo` + `file:/`. Also, I don't like the fact that `RemoteRepo` in the code now means "a remote repo, except when it's not" :-).

> This would also require a further change to the config file. 

I think this'd be for the better from the UX point of view, as `local-repo` and `remote-repo` would be nicely symmetric.